openedx/core/djangoapps/call_stack_manager/core.py

```diff
@@ -23,122 +27,207 @@
 
-    Args:
-        current_model - Name of the model class
     """
-    # holds temporary callstack
-    # frame[0][6:-1] -> File name along with path
-    # frame[1][6:] -> Line Number
-    # frame[2][3:] -> Context
-    temp_call_stack = [(frame[0][6:-1],
-                        frame[1][6:],
-                        frame[2][3:])
-                       for frame in [stack.replace("\n", "").strip().split(',') for stack in traceback.format_stack()]
+    # Holds temporary callstack
+    # List with each element 4-tuple(filename, line number, function name, text)
+    # and filtered with respect to regular expressions
+    temp_call_stack = [frame for frame in [frames for frames in traceback.extract_stack()]
                        if not any(reg.match(frame[0]) for reg in REGULAR_EXPS)]
 
-    # avoid duplication.
-    if temp_call_stack not in STACK_BOOK[current_model] and TRACK_FLAG \
-            and not issubclass(current_model, tuple(HALT_TRACKING)):
-        STACK_BOOK[current_model].append(temp_call_stack)
-        log.info("logging new call stack for %s:\n %s", current_model, temp_call_stack)
+    def _print(frame):
+        """ Converts the tuple format of frame to a printable format
+
+        Arguments:
+            frame - current frame tuple of format (file, line number, function name, context)
+
+        Returns:
+            frame converted to a string in following format
+            File XXX, line number XXX, in XXX
+                XXXXX
+        """
+        return str('\n File ' + str(frame[0]) + ', line number ' + str(frame[1]) + ', in ' +
+                   str(frame[2]) + '\n\t' + str(frame[3]))
+
+    # Customize output of call stack
+    final_call_stack = ""
+    for frame in temp_call_stack:
+        final_call_stack += _print(frame)
+
+    def _should_get_logged(entity_name):
+        """ checks if current call stack of current entity should be logged or not
+
+        Arguments:
+            entity_name - Name of the current entity
+
+        Returns:
+            True if the current call stack is to logged, False otherwise
+        """
+        if HALT_TRACKING:
+            if not HALT_TRACKING[-1]: # if it is None
+                return False
+            else:
+                if temp_call_stack in STACK_BOOK[entity_name]:
+                    return False
+                else:
+                    if inspect.isclass(entity_name):
+                        if not issubclass(entity_name, tuple(HALT_TRACKING[-1])):
+                            return True
+                        else:
+                            return False
+                    else:
+                        if entity_name not in tuple(HALT_TRACKING[-1]):
+                            return True
+                        else:
+                            return False
+        else:
+            if temp_call_stack in STACK_BOOK[entity_name]:
+                return False
+            else:
+                return True
```

Comment:
It looks like if `temp_call_stack in STACK_BOOK[entity_name]`, then we always return False, so you could just do that check first, and get rid of the duplicate conditions.